### PR TITLE
Vectorize PGO

### PR DESCRIPTION
**Pull request by luisenp**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by luisenp — COMMENTED** [review]
Left some comments for discussion. Depending on what we decide we can either merge as is, or make some refactoring before merging.

**Comment by luisenp** [comment]
@fantaosha Left some comments above, just noticed that I created this PR, so not sure if you got notifications.

File: examples/configs/pose_graph_benchmark.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+dataset: sphere2500
+hydra:
+  run:
+    dir: examples/outputs/pgo/benchmark
+  sweep:
+    dir: examples/outputs/pgo/benchmark
```
Comment (mhmukadam) [inline examples/configs/pose_graph_benchmark.yaml:6]: missing new line

File: examples/configs/pose_graph_cube.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  max_iters: 10 
+  step_size: 1
+  verbose: true
+  reg_w: 1e-3
+
+hydra:
+  run:
+    dir: examples/outputs/pgo/pose_graph_cube
+  sweep:
+    dir: examples/outputs/pgo/pose_graph_cube
```
Comment (mhmukadam) [inline examples/configs/pose_graph_cube.yaml:19]: missing new line

File: examples/pose_graph_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
[the remark this answers is not in the document]
Comment (mhmukadam) [reply examples/pose_graph_benchmark.py:1]: Let's also add some comments throughout these examples (similar to other examples in this folder) and update the example/REAMDE.md appropriately.

File: theseus/utils/examples/__init__.py [hunk 4]
```diff
@@ -50,4 +50,10 @@
     )
 
 
-from .pose_graph import PosePriorError, RelativePoseError
+from .pose_graph import (
+    PoseGraphDataset,
+    PoseGraphEdge,
+    PosePriorError,
+    RelativePoseError,
```
Comment (mhmukadam) [inline theseus/utils/examples/__init__.py:57]: We can get rid of these (and their files) now since we have moved to using Difference and Between, right?

**Review by mhmukadam — COMMENTED** [review]
Looks good, left some comments related to documentation and cleanup.

File: theseus/utils/examples/__init__.py (hunk 4 above)
Comment (fantaosha) [reply theseus/utils/examples/__init__.py:57]: Yes, they are equivalent to `Difference` and `Between`.

**Review by luisenp — COMMENTED** [review]
LGTM, only some minor comments. @fantaosha

File: examples/pose_graph_synthetic.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    pose_indices: List[int] = [index for index, _ in enumerate(pg_batch.poses)]
+    gt_pose_indices: List[int] = []
+
+    forward_times = []
+    backward_times = []
+    forward_mems = []
+    backward_mems = []
+
+    for edge in pg_batch.edges:
+        relative_pose_cost = th.eb.Between(
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply examples/pose_graph_synthetic.py]: Fixed

**Pull request merged**